packages/react-components/react-tree/src/components/TreeItem/useTreeItem.tsx

```diff
@@ -192,16 +193,42 @@ export function useTreeItem_unstable(props: TreeItemProps, ref: React.Ref<HTMLDi
     }
   });
 
-  const handleActionsVisible = useEventCallback((event: React.FocusEvent | React.MouseEvent) => {
+  const setActionsVisibleIfNotFromSubtree = React.useCallback((event: React.SyntheticEvent<HTMLDivElement>) => {
     const isTargetFromSubtree = Boolean(
       subtreeRef.current && elementContains(subtreeRef.current, event.target as Node),
     );
     if (!isTargetFromSubtree) {
       setActionsVisible(true);
     }
+  }, []);
+
+  const handleMouseOver = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
+    onMouseOver?.(event);
+    setActionsVisibleIfNotFromSubtree(event);
+  });
+
+  const handleFocus = useEventCallback((event: React.FocusEvent<HTMLDivElement>) => {
+    onFocus?.(event);
+    setActionsVisibleIfNotFromSubtree(event);
   });
 
-  const handleActionsInvisible = useEventCallback((event: React.FocusEvent | React.MouseEvent) => {
+  const handleMouseOut = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
+    onMouseOut?.(event);
+    const isTargetFromSubtree = Boolean(
```

Comment:
you can hoist the related target logic too, since it's used by both mouseout and blur